error: cannot convert 'struct _option_string' to 'unsigned char *'

### Describe the bug

1) "This is a compiler bug, please report it using `v bug file.v`."

2) the error messages I'm getting aren't helping me understand what I did wrong (see below)

3) it's not giving me .v file line numbers, but .tmp.c line numbers.

```
C:/Users/royce3/AppData/Local/Temp/v_0/bug.6399603458033988116.tmp.c:7638: warning: cast between pointer and integer of different size
C:/Users/royce3/AppData/Local/Temp/v_0/bug.6399603458033988116.tmp.c:12627: error: cannot convert 'struct _option_string' to 'unsigned char *'
```


### Expected Behavior

Get a compilation error telling me what I did wrong and where.

### Current Behavior

```
bug.v:7:2: warning: unused variable: `password2`
    5 |
    6 | pub fn connect( params ConnectParams ) {
    7 |     password2 := ?string( params.password )
      |     ~~~~~~~~~
    8 | }
==================
C:/Users/royce3/AppData/Local/Temp/v_0/bug.1434562961698783307.tmp.c:7638: warning: cast between pointer and integer of different size
C:/Users/royce3/AppData/Local/Temp/v_0/bug.1434562961698783307.tmp.c:12627: error: cannot convert 'struct _option_string' to 'unsigned char *'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```


### Reproduction Steps

```
[params]
struct ConnectParams {
	password ?string
}

pub fn connect( params ConnectParams ) {
	password2 := ?string( params.password )
}
```


### Possible Solution

no idea, sorry.

### Additional Information/Context

I'm porting some python code and it stores the supplied password in it's __init__ method so I thought I'd store it as a local variable until I'm ready to create the Connection object. In hindsight I won't need to do this, but there will be many scenario where I need to make a local copy of an object because I need to apply some overrides to it's value depending on the scenario.

### V version

V 0.3.4 0fc33c6

### Environment details (OS name and version, etc.)

V full version: V 0.3.4 4dc3388.0fc33c6
OS: windows, Microsoft Windows 10 Home v19045 64-bit
Processor: 4 cpus, 64bit, little endian,

getwd: C:\cvs\pg
vexe: C:\cvs\v\v.exe
vexe mtime: 2023-06-02 12:55:43

vroot: OK, value: C:\cvs\v
VMODULES: OK, value: C:\Users\royce3\.vmodules
VTMP: OK, value: C:\Users\royce3\AppData\Local\Temp\v_0

Git version: git version 2.37.3.windows.1
Git vroot status: weekly.2023.22-5-g0fc33c6f
.git/config present: true

CC version: Error: exec failed (CreateProcess) with code 2: The system cannot find the file specified.
 cmd: cc --version
thirdparty/tcc status: thirdparty-windows-amd64 e90c2620

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -93,7 +93,7 @@ fn (mut g Gen) expr_with_opt(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type)
 		g.inside_opt_or_res = old_inside_opt_or_res
 	}
 	if expr_typ.has_flag(.option) && ret_typ.has_flag(.option)
-		&& expr in [ast.DumpExpr, ast.Ident, ast.ComptimeSelector, ast.AsCast, ast.CallExpr, ast.MatchExpr, ast.IfExpr, ast.IndexExpr, ast.UnsafeExpr, ast.CastExpr] {
+		&& expr in [ast.SelectorExpr, ast.DumpExpr, ast.Ident, ast.ComptimeSelector, ast.AsCast, ast.CallExpr, ast.MatchExpr, ast.IfExpr, ast.IndexExpr, ast.UnsafeExpr, ast.CastExpr] {
 		if expr in [ast.Ident, ast.CastExpr] {
 			if expr_typ.idx() != ret_typ.idx() {
 				return g.expr_opt_with_cast(expr, expr_typ, ret_typ)

--- a/vlib/v/tests/option_selector_cast_test.v
+++ b/vlib/v/tests/option_selector_cast_test.v
@@ -1,0 +1,15 @@
+[params]
+struct ConnectParams {
+	password ?string
+}
+
+pub fn connect(params ConnectParams) ? {
+	password2 := ?string(params.password)
+	assert password2? == 'foo'
+}
+
+fn test_main() {
+	connect(ConnectParams{
+		password: 'foo'
+	})
+}

